python/lsst/scarlet/lite/image.py

```diff
@@ -0,0 +1,1372 @@
+    in an image. For example:
+
+    >>> print(image[("r", "g", "i")])
+    Image:
+     [[[20 21 22 23 24]
+      [25 26 27 28 29]
+      [30 31 32 33 34]
+      [35 36 37 38 39]]
+    <BLANKLINE>
+     [[ 0  1  2  3  4]
+      [ 5  6  7  8  9]
+      [10 11 12 13 14]
+      [15 16 17 18 19]]
+    <BLANKLINE>
+     [[40 41 42 43 44]
+      [45 46 47 48 49]
+      [50 51 52 53 54]
+      [55 56 57 58 59]]]
+      bands=('r', 'g', 'i')
+      bbox=<Box shape=(4, 5), origin=(20, 30)>
+
+
+    will return a new image with the bands re-ordered.
+
+    Images can be combined using the standard arithmetic operations similar to
+    numpy arrays, including ``+, -, *, /, **`` etc, however, if two images are
+    combined with different bounding boxes, the _union_ of the two
+    boxes is used for the result. For example:
+
+    >>> image1 = Image(np.ones((2, 3, 4)), bands=tuple("gr"))
+    >>> image2 = Image(np.ones((2, 3, 4)), bands=tuple("gr"), yx0=(2, 3))
+    >>> result = image1 + image2
+    >>> print(result)
+    Image:
+     [[[1. 1. 1. 1. 0. 0. 0.]
+      [1. 1. 1. 1. 0. 0. 0.]
+      [1. 1. 1. 2. 1. 1. 1.]
+      [0. 0. 0. 1. 1. 1. 1.]
+      [0. 0. 0. 1. 1. 1. 1.]]
+    <BLANKLINE>
+     [[1. 1. 1. 1. 0. 0. 0.]
+      [1. 1. 1. 1. 0. 0. 0.]
+      [1. 1. 1. 2. 1. 1. 1.]
+      [0. 0. 0. 1. 1. 1. 1.]
+      [0. 0. 0. 1. 1. 1. 1.]]]
+      bands=('g', 'r')
+      bbox=<Box shape=(5, 7), origin=(0, 0)>
+
+    If instead you want to additively ``insert`` image 1 into image 2,
+    so that they have the same bounding box as image 2, use
+
+    >>> _ = image2.insert(image1)
+    >>> print(image2)
+    Image:
+     [[[2. 1. 1. 1.]
+      [1. 1. 1. 1.]
+      [1. 1. 1. 1.]]
+    <BLANKLINE>
+     [[2. 1. 1. 1.]
+      [1. 1. 1. 1.]
+      [1. 1. 1. 1.]]]
+      bands=('g', 'r')
+      bbox=<Box shape=(3, 4), origin=(2, 3)>
+
+    To insert an image using a different operation use
+
+    >>> from operator import truediv
+    >>> _ = image2.insert(image1, truediv)
+    >>> print(image2)
+    Image:
+     [[[2. 1. 1. 1.]
+      [1. 1. 1. 1.]
+      [1. 1. 1. 1.]]
+    <BLANKLINE>
+     [[2. 1. 1. 1.]
+      [1. 1. 1. 1.]
+      [1. 1. 1. 1.]]]
+      bands=('g', 'r')
+      bbox=<Box shape=(3, 4), origin=(2, 3)>
+
+
+    However, depending on te operation you may get unexpected results
+    since now there could be ``NaN`` and ``inf`` values due to the zeros
+    in the non-overlaping regions.
+    Instead, to select only the overlap region one can use
+
+    >>> result = image1 / image2
+    >>> print(result[image1.bbox & image2.bbox])
+    Image:
+     [[[0.5]]
+    <BLANKLINE>
+     [[0.5]]]
+      bands=('g', 'r')
+      bbox=<Box shape=(1, 1), origin=(2, 3)>
+
+
+    Parameters
+    ----------
+    data:
+        The array data for the image.
+    bands:
+        The bands coving the image.
+    yx0:
+        The (y, x) offset for the lower left of the image.
+    indices:
+        Dictionary of cached indices with the bands to project this image
```

Comment:
I can probably remove this. This was basically to make it so that combining images (like source models) would only have to calculate the projected bounding box once as opposed to doing it in each iteration. But in reality this is an inexpensive operation and probably isn't worth the (slight) memory overhead.